packages/status-js/src/client/community/community.ts

```diff
@@ -85,380 +66,125 @@ export class Community {
 
-  private observeCommunity = () => {
-    this.waku.relay.addDecryptionKey(this.communityDecryptionKey)
-    this.waku.relay.addObserver(this.client.handleWakuMessage, [
-      this.communityContentTopic,
-    ])
+    this.client.waku.relay.addObserver(
+      this.client.handleWakuMessage,
+      contentTopics
+    )
   }
 
-  private observeChannelMessages = async (chatsIds: string[]) => {
-    const symKeyPromises = chatsIds.map(async (chatId: string) => {
-      const id = `${this.communityPublicKey}${chatId}`
-      const channelContentTopic = idToContentTopic(id)
-
-      const symKey = await createSymKeyFromPassword(id)
+  private unobserveChatMessages = (
+    chatDescription: CommunityDescription['chats']
+  ) => {
+    const contentTopics = Object.keys(chatDescription).map(chatUuid => {
+      const chat = this.chats.get(chatUuid)
+      const contentTopic = chat!.contentTopic
 
-      this.waku.relay.addDecryptionKey(symKey, {
-        method: waku_message.DecryptionMethod.Symmetric,
-        contentTopics: [channelContentTopic],
-      })
+      this.chats.delete(chatUuid)
 
-      return channelContentTopic
+      return contentTopic
     })
-    const contentTopics = await Promise.all(symKeyPromises)
-
-    this.waku.relay.addObserver(this.client.handleWakuMessage, contentTopics)
-  }
 
-  private unobserveChannelMessages = (chatIds: string[]) => {
-    const contentTopics = createChannelContentTopics(
-      chatIds,
-      this.communityPublicKey
+    this.client.waku.relay.deleteObserver(
+      this.client.handleWakuMessage,
+      contentTopics
     )
-
-    this.waku.relay.deleteObserver(this.client.handleWakuMessage, contentTopics)
   }
 
-  public handleCommunityMetadataEvent = (
-    communityMetadata: CommunityMetadataType
-  ) => {
-    if (this.communityMetadata) {
-      if (this.communityMetadata.clock > communityMetadata.clock) {
+  public handleChange = (description: CommunityDescription) => {
+    if (this.description) {
+      // already handled
+      if (this.description.clock >= description.clock) {
         return
       }
 
-      // Channels
-      const removedChats = difference(
-        Object.keys(this.communityMetadata.chats),
-        Object.keys(communityMetadata.chats)
-      )
-      const addedChats = difference(
-        Object.keys(communityMetadata.chats),
-        Object.keys(this.communityMetadata.chats)
+      // Chats
+      // observe
+      const removedChats = getDifferenceByKeys(
+        this.description.chats,
+        description.chats
       )
-
       if (removedChats.length) {
-        this.unobserveChannelMessages(removedChats)
+        this.unobserveChatMessages(removedChats)
       }
 
+      const addedChats = getDifferenceByKeys(
+        description.chats,
+        this.description.chats
+      )
       if (addedChats.length) {
-        this.observeChannelMessages(addedChats)
+        this.observeChatMessages(addedChats)
       }
```

Comment:
`getDifferenceByKeys` currently returns an object, which does not have `length` property.

Reply:
Great catch! Didn't test that, thanks.